exporter/exporterhelper/internal/queuebatch/persistent_queue.go

```diff
@@ -191,6 +204,52 @@ func (pq *persistentQueue[T]) restoreQueueSizeFromStorage(ctx context.Context) (
 	return bytesToItemIndex(val)
 }
 
+// marshalCurrentMetadata constructs metadata from current pq state and returns the marshaled bytes.
+// This is a helper and does not perform the Set operation itself.
+// pq.mu must be held.
+// TODO: Enable when https://github.com/open-telemetry/opentelemetry-collector/issues/12890 is done
+//
+//nolint:unused
+func (pq *persistentQueue[T]) marshalCurrentMetadata() ([]byte, error) {
+	meta := queueMetadata{
+		ReadIndex:                pq.readIndex,
+		WriteIndex:               pq.writeIndex,
+		CurrentlyDispatchedItems: pq.currentlyDispatchedItems,
+	}
+	if !pq.isRequestSized {
+		meta.QueueSize = pq.queueSize
+	} else {
+		//nolint:gosec
+		calculatedSize := int64(pq.writeIndex - pq.readIndex)
+		if calculatedSize < 0 {
+			calculatedSize = 0
+		}
+		meta.QueueSize = calculatedSize
```

Comment:
The size should always be "correctly" calculated, unless I am missing something.